File: server/src/lessonState.ts
import type { LessonState } from "@prisma/client";

/**
 * Wire shape consumed by the frontend's ProgressContext — deliberately the
 * same LessonProgress object the UI already worked with, so no lesson/stage
 * component had to change when this moved from localStorage to the database.
 */
export interface LessonProgressDTO {
  lessonId: string;
  completedStages: string[];
  vocabIndex: number;
  miniTestResult?: { correct: number; total: number; completedAt: string };
  practiceResult?: { correct: number; total: number; completedAt: string };
  reviewResult?: { correct: number; total: number; completedAt: string };
  startedAt: string;
  completedAt?: string;
}

export function toDTO(state: LessonState): LessonProgressDTO {
  const dto: LessonProgressDTO = {
    lessonId: state.lessonId,
    completedStages: state.completedStages,
    vocabIndex: state.vocabIndex,
    startedAt: state.startedAt.toISOString(),
  };
  if (state.completedAt) dto.completedAt = state.completedAt.toISOString();
  if (state.miniTestTotal !== null && state.miniTestCorrect !== null) {
    dto.miniTestResult = {
      correct: state.miniTestCorrect,
      total: state.miniTestTotal,
      completedAt: (state.miniTestAt ?? state.updatedAt).toISOString(),
    };
  }
  if (state.practiceTotal !== null && state.practiceCorrect !== null) {
    dto.practiceResult = {
      correct: state.practiceCorrect,
      total: state.practiceTotal,
      completedAt: (state.practiceAt ?? state.updatedAt).toISOString(),
    };
  }
  if (state.reviewTotal !== null && state.reviewCorrect !== null) {
    dto.reviewResult = {
      correct: state.reviewCorrect,
      total: state.reviewTotal,
      completedAt: (state.reviewAt ?? state.updatedAt).toISOString(),
    };
  }
  return dto;
}

File: backend/app/services/lesson_state.py
from app.models.lesson_state import LessonState
from app.utils import to_iso_z


def to_dto(state: LessonState) -> dict:
    """Wire shape consumed by the (future) Flutter client's progress store —
    deliberately the same LessonProgress object the old React UI worked
    with, mirroring lessonState.ts's toDTO exactly."""
    dto: dict = {
        "lessonId": state.lessonId,
        "completedStages": state.completedStages,
        "vocabIndex": state.vocabIndex,
        "startedAt": to_iso_z(state.startedAt),
    }
    if state.completedAt:
        dto["completedAt"] = to_iso_z(state.completedAt)
    if state.miniTestTotal is not None and state.miniTestCorrect is not None:
        dto["miniTestResult"] = {
            "correct": state.miniTestCorrect,
            "total": state.miniTestTotal,
            "completedAt": to_iso_z(state.miniTestAt or state.updatedAt),
        }
    if state.practiceTotal is not None and state.practiceCorrect is not None:
        dto["practiceResult"] = {
            "correct": state.practiceCorrect,
            "total": state.practiceTotal,
            "completedAt": to_iso_z(state.practiceAt or state.updatedAt),
        }
    if state.reviewTotal is not None and state.reviewCorrect is not None:
        dto["reviewResult"] = {
            "correct": state.reviewCorrect,
            "total": state.reviewTotal,
            "completedAt": to_iso_z(state.reviewAt or state.updatedAt),
        }
    if state.nodeResults is not None:
        dto["nodeResults"] = state.nodeResults
    return dto
